### ISSUE TITLE
SIGSEGV when using SKCanvas.DrawText() and then SKCanvas.Flush() 

### ISSUE THREAD

**Issue by mattleibow** [mono/SkiaSharp#647]
### Description

The app crashes when drawing text and then flushing the canvas. Waiting on a response from Google to address this.

### Basic Information

- Version with issue:  1.68.0
- Last known good version:  1.60.3
- IDE:  N/A
- Platform Target Frameworks:
  - Android: 8.1 
- Target Devices:
  -   Pixel 2 and Razer Phone

### Links

 - skia bug: https://bugs.chromium.org/p/skia/issues/detail?id=8425
 - "repro" case: https://github.com/mattleibow/NativeAndroidSkiaSharp

> VS bug [#736300](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/736300)

**Comment by mattleibow** [comment]
After doing some more research, I found out several things:
 - it works fine in the emulator
 - it works fine as a debug build

As a result, there appears to be something going wrong for ARM when the optimizations are at `-Os`. I had a try with `-O3` and it appears to solve the issue.

However, this does incur a penalty with the size no longer being the main point. As a result, the size increases by about 800KB-12KB depending on the architecture.

For this release, I will go with a larger binary that works over a smaller one that can't do GPU. I have reported the issue to Google along with all the information I could find. Hopefully, they can find out exactly what the issue is and provide a real fix.

**Comment by apaatsio** [comment]
I found this issue on google because I'm experiencing a similar bug on a Flutter app. See https://github.com/flutter/flutter/issues/29751 and [my comment](https://github.com/flutter/flutter/issues/29751#issuecomment-478231119)

There seem to be many common denominators (Pixel 2, Android 8.1, drawing text, SIGSEGV during SKCanvas.Flush()) so I was wondering if these things could be related.

Unfortunately, I can't see the Skia issue 8425 that @mattleibow linked. It's giving me "Permission denied" so I can't dig deeper into this.

**Comment by blimkemann** [comment]
I am getting this error even in debug builds using 1.68.0 and above when targeting Android 10 as a build target. It happens every time and is repeatable. If I remove all calls to DrawText there is no crash. Even a single DrawText will cause a crash with a "Could not allocate vertices" message. 

Is there any news on this?

**Comment by mattleibow** [comment]
@apaatsio, I had a look at that issue on the Google bug list, no feedback at all. The only comment a long while back was:

> Well, it is dying while trying to upload glyphs to the font atlas.
>
> I don't know why that would be occurring though.

@blimkemann could you create a test/repo project for your example. It was working when I reduced the optimization level a while back. Maybe this is a slightly different case that I can start testing on.

> I am looking on getting a brand new skia engine out in #986, just trying to fix the last few bugs in the current version. Moving this issue to that release so it gets at least a test.

**Comment by blimkemann** [comment]
[SkiaTextBug.zip](https://github.com/mono/SkiaSharp/files/4180786/SkiaTextBug.zip)
@mattleibow, here is a repro. Pretty simple really. Just start in debugger on Android and the drawing will crash as soon as Canvas.Flush is called. Please let me know if you have questions about this.

**Comment by mattleibow** [comment]
Thanks! I'll have a look.

**Comment by mattleibow** [comment]
This has been some time, I will try using the new v2.80 packages. Have you still been having this issue?

**Comment by blimkemann** [comment]
Yes, still having this issue.

**Comment by mattleibow** [comment]
Not sure how I missed it, but the issue with your code is that you are drawing outside of the main render loop.

This code:

```csharp
MainThread.BeginInvokeOnMainThread(() => {
    // drawing code
});
```

The `MainThread.BeginInvokeOnMainThread` returns immediately and the surface/GL context is reset.

Just remove those blocks and draw directly.

<img src="https://user-images.githubusercontent.com/1096616/85861265-596ad980-b7c0-11ea-92cb-ba5277d45e57.png" width="300" />

**Comment by blimkemann** [comment]
Thanks. That fixed that issue. I am now always wrapping InvalidateSurface in MainThread.BeginInvokeOnMainThread.

However, as I use some commands on the page and drawing happens, after about 4 calls to InvalidateSurface, the whole SKGLView appears as a black rectangle (only on Android) and then randomly comes back after a few more redraws. Am I still missing something?

**Comment by mattleibow** [comment]
Are there any errors in the log? Is there something you are doing just before it happens?

**Comment by blimkemann** [comment]
Each time, I call Canvas.Clear(), Canvas.Flush(), do my drawing, and Canvas.Flush() again. Other than that, exactly the same thing happens between drawings.

The logs show only this:
Time
Device Name
Type
PID
Tag
Message
07-02 11:26:07.704
moto_g7
Info
24845
.redbook.mobil
Explicit concurrent copying GC freed 10877(367KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 3MB/6MB, paused 311us total 7.868ms
07-02 11:26:07.707
moto_g7
Debug
24845
Mono
GC_TAR_BRIDGE bridges 7 objects 95968 opaque 280479 colors 7 colors-bridged 7 colors-visible 7 xref 0 cache-hit 0 cache-semihit 0 cache-miss 0 setup 0.02ms tarjan 16.60ms scc-setup 0.02ms gather-xref 0.01ms xref-setup 0.01ms cleanup 2.08ms
07-02 11:26:07.707
moto_g7
Debug
24845
Mono
GC_BRIDGE: Complete, was running for 10.88ms
07-02 11:26:07.707
moto_g7
Debug
24845
Mono
GC_MINOR: (Nursery full) time 20.11ms, stw 21.25ms promoted 676K major size: 55024K in use: 50942K los size: 4096K in use: 2560K

**Comment by mattleibow** [comment]
@blimkemann Is this still happening? Have you tried the v2 of SKiaSharp?

**Comment by blimkemann** [comment]
It is not still happening. Turns out is was a threading issue

**Comment by mattleibow** [comment]
Thanks for the feedback. I'm closing it now. If things start happening, then feel free to open a new issue and link to this for context.
